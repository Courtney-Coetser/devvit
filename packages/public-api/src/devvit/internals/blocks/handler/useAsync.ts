import type { JSONValue } from '@devvit/shared-types/json.js';
import type { AsyncUseStateInitializer, UseAsyncResult } from '../../../../types/hooks.js';

import type { AsyncResponse, UIEvent } from '@devvit/protos';
import { CIRCUIT_BREAKER_MSG } from '@devvit/shared-types/CircuitBreaker.js';
import isEqual from 'lodash.isequal';
import { registerHook } from './BlocksHandler.js';
import type { RenderContext } from './RenderContext.js';
import type { Hook, HookParams } from './types.js';
export type AsyncOptions = {
  /**
   * The data loader will re-run if the value of `depends` changes.
   */
  depends?: JSONValue;
};

export type LoadState = 'initial' | 'loading' | 'loaded' | 'error' | 'disabled';

type AsyncState<S extends JSONValue> = {
  depends: JSONValue;
  load_state: LoadState;
  data: S | null;
  error: { message: string; details: string } | null;
};

class AsyncHook<S extends JSONValue> implements Hook {
  state: AsyncState<S>;
  readonly #debug: boolean;
  #hookId: string;
  #initializer: AsyncUseStateInitializer<S>;
  #invalidate: () => void;
  #ctx: RenderContext;
  #localDepends: JSONValue;

  constructor(initializer: AsyncUseStateInitializer<S>, options: AsyncOptions, params: HookParams) {
    this.#debug = !!params.context.devvitContext.debug.useAsync;
    if (this.#debug) console.debug('[useAsync] v1', options);
    this.state = { data: null, load_state: 'initial', error: null, depends: null };
    this.#hookId = params.hookId;
    this.#initializer = initializer;
    this.#invalidate = params.invalidate;
    this.#ctx = params.context;
    this.#localDepends = options.depends ?? null;
  }

  /**
   * After we look at our state, we need to decide if we need to dispatch a request to load the data.
   */
  onStateLoaded(): void {
    if (this.state.load_state === 'disabled') {
      return;
    }
    if (this.#debug) console.debug('[useAsync] async onLoad ', this.#hookId, this.state);
    if (this.#debug)
      console.debug('[useAsync] async onLoad have ', this.#localDepends, 'and', this.state.depends);
    if (!isEqual(this.#localDepends, this.state.depends) || this.state.load_state === 'initial') {
      this.state.load_state = 'loading';
      this.state.depends = this.#localDepends;
      this.#invalidate();

      const requeueEvent: UIEvent = {
        hook: this.#hookId,
        async: true,
        asyncRequest: {
          requestId: this.#hookId + '-' + JSON.stringify(this.state.depends),
        },
      };
      if (this.#debug) console.debug('[useAsync] onLoad requeue');
      this.#ctx.addToRequeueEvents(requeueEvent);
    }
  }

  async onUIEvent(event: UIEvent, context: RenderContext): Promise<void> {
    /**
     * Requests and responses are both handled here. If we have a request, we need to load the data
     * and then send a response.  If we have a response, we need to update our state and re-render.
     *
     * This is a very event-driven way to handle state, but it's the only way to handle async state.
     */
    if (event.asyncRequest) {
      const asyncResponse: AsyncResponse = { requestId: event.asyncRequest.requestId };
      try {
        asyncResponse.data = {
          value: await this.#initializer(),
        };
      } catch (e) {
        if (e instanceof Error) {
          if (e.message === CIRCUIT_BREAKER_MSG) {
            throw e;
          }
          asyncResponse.error = { message: e.message, details: e.stack ?? '' };
        } else {
          asyncResponse.error = { message: 'Unknown error', details: String(e) };
        }
      }

      const requeueEvent: UIEvent = {
        asyncResponse: asyncResponse,
        hook: this.#hookId,
      };
      if (this.#debug) console.debug('[useAsync] onReq requeue');
      context.addToRequeueEvents(requeueEvent);
    } else if (event.asyncResponse) {
      const anticipatedRequestId = this.#hookId + '-' + JSON.stringify(this.state.depends);
      if (event.asyncResponse.requestId === anticipatedRequestId) {
        this.state = {
          ...this.state,
          data: event.asyncResponse.data?.value as S | null,
          error: event.asyncResponse.error ?? null,
          load_state: event.asyncResponse.error ? 'error' : 'loaded',
        };
        this.#invalidate();
      } else {
        if (this.#debug) console.debug('[useAsync] onResp skip, stale event');
      }
    } else {
      throw new Error('Unknown event type');
    }
  }
}

/**
 * This is the preferred way to handle async state in Devvit.
 *
 * @param initializer -- any async function that returns a JSONValue
 * @returns UseAsyncResult<S>
 */
export function useAsync<S extends JSONValue>(
  initializer: AsyncUseStateInitializer<S>,
  options: AsyncOptions = {}
): UseAsyncResult<S> {
  const hook = registerHook({ namespace: 'useAsync' }, (params) => {
    return new AsyncHook(initializer, options, params);
  });

  return {
    data: hook.state.data,
    error: hook.state.error,
    loading: hook.state.load_state === 'loading',
  };
}
