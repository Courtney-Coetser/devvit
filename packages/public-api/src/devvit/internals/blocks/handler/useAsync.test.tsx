/** @jsx Devvit.createElement */
/** @jsxFrag Devvit.Fragment */

import type { UIEvent, UIRequest } from '@devvit/protos';
import { Devvit, useState } from '../../../../index.js';
import type { JSONValue } from '@devvit/shared-types/json.js';
import { useAsync } from './useAsync.js';
import type { HookRef } from './types.js';
import { captureHookRef } from './refs.js';
import { BlocksHandler } from './BlocksHandler.js';
import { EmptyRequest, findHookState, mockMetadata } from './test-helpers.js';
import { CircuitBreak } from '@devvit/shared-types/CircuitBreaker.js';
let shouldThrow = false;

const loader = async (): Promise<string> => {
  if (shouldThrow) {
    throw CircuitBreak('');
  }
  await setTimeout(() => {}, 10);
  return 'hi world!';
};

const asyncRef: HookRef = {};
const buttonRef: HookRef = {};

const App: Devvit.BlockComponent = () => {
  const [depends, setDepends] = useState(1);
  const { data, error, loading } = captureHookRef(useAsync(loader, { depends }), asyncRef);
  return (
    <hstack>
      <text>{loading ? 'loading' : data || (error ?? 'none').toString()}</text>
      <button onPress={captureHookRef(() => setDepends((d) => d + 1), buttonRef)} />
    </hstack>
  );
};

const NullApp: Devvit.BlockComponent = () => {
  const { data, error, loading } = captureHookRef(
    useAsync(async () => null, { depends: 1 }),
    asyncRef
  );
  return <text>{loading ? 'loading' : data || (error ?? 'none').toString()}</text>;
};

const asyncRequestEvent = (ref: HookRef, depends: JSONValue = 1): UIEvent => {
  return {
    asyncRequest: { requestId: (ref.id ?? '') + '-' + JSON.stringify(depends) },
    hook: ref.id,
    async: true,
  };
};

const asyncResponseEvent = (ref: HookRef, depends: JSONValue = 1): UIEvent => {
  return {
    asyncResponse: {
      requestId: (ref.id ?? '') + '-' + JSON.stringify(depends),
      data: { value: 'hi world!' },
      error: undefined,
    },
    hook: ref.id,
  };
};
describe('regressions', () => {
  test('loading null doesnt loop', async () => {
    shouldThrow = true;
    const handler = new BlocksHandler(NullApp);

    let response = await handler.handle(EmptyRequest, mockMetadata);
    expect(JSON.stringify(response.blocks)).toContain('loading');
    expect(response.state).toMatchInlineSnapshot(`
      {
        "NullApp.useAsync-0": {
          "data": null,
          "depends": 1,
          "error": null,
          "load_state": "loading",
        },
        "__cache": {},
      }
    `);
    const event = {
      asyncResponse: {
        requestId: (asyncRef.id ?? '') + '-1',
        data: { value: null },
        error: undefined,
      },
      hook: asyncRef.id,
    };

    const request: UIRequest = { events: [event], state: response.state };
    response = await handler.handle(request, mockMetadata);

    expect(response.state).toMatchInlineSnapshot(`
      {
        "NullApp.useAsync-0": {
          "data": null,
          "depends": 1,
          "error": null,
          "load_state": "loaded",
        },
        "__cache": {},
      }
    `);
    expect(findHookState(asyncRef)).toMatchSnapshot();
    expect(JSON.stringify(response.blocks)).toContain('none');
    expect(response.effects.length).toEqual(0);
    expect(response.events.length).toEqual(0);
  });

  test('event-driven state changes are reflected', async () => {
    const handler = new BlocksHandler(App);
    let request: UIRequest = EmptyRequest;
    let response = await handler.handle(request, mockMetadata);

    request = {
      events: [
        {
          userAction: { actionId: buttonRef.id ?? 'wtf' },
          hook: buttonRef.id ?? 'wtf',
        },
      ],
      state: response.state,
    };
    response = await handler.handle(request, mockMetadata);
    expect(response.state!['App.useState-0'].value).toEqual(2);
  });

  test("error response doesn't requeue", async () => {
    const handler = new BlocksHandler(App);
    let request: UIRequest = EmptyRequest;
    let response = await handler.handle(request, mockMetadata);

    const rspEvent = asyncResponseEvent(asyncRef);
    rspEvent.asyncResponse!.error = { message: 'some other error', details: 'some details' };
    rspEvent.asyncResponse!.data = undefined;

    request = { events: [rspEvent], state: response.state };

    response = await handler.handle(request, mockMetadata);
    expect(response.state).toMatchInlineSnapshot(`
      {
        "App.useAsync-1": {
          "data": undefined,
          "depends": 1,
          "error": {
            "details": "some details",
            "message": "some other error",
          },
          "load_state": "error",
        },
        "__cache": {},
      }
    `);
    expect(response.events.length).toEqual(0);
  });

  test('multiple useAsyncs can resolve when they are dependent', async () => {
    shouldThrow = true;
    const async1Ref: HookRef = {};
    const async2Ref: HookRef = {};
    const AppUseAsyncs: Devvit.BlockComponent = () => {
      const {
        data: data1,
        error: error1,
        loading: loading1,
      } = captureHookRef(useAsync(loader), async1Ref);
      const {
        data: data2,
        error: error2,
        loading: loading2,
      } = captureHookRef(useAsync(loader, { depends: data1 }), async2Ref);
      return (
        <hstack>
          <text>{loading1 ? 'loading' : data1 || (error1 ?? 'none').toString()}</text>
          <text>{loading2 ? 'loading' : data2 || (error2 ?? 'none').toString()}</text>
        </hstack>
      );
    };

    const handler = new BlocksHandler(AppUseAsyncs);
    const request1: UIRequest = EmptyRequest;
    const response1 = await handler.handle(request1, mockMetadata);

    expect(response1.state).toMatchInlineSnapshot(`
      {
        "AppUseAsyncs.useAsync-0": {
          "data": null,
          "depends": null,
          "error": null,
          "load_state": "loading",
        },
        "AppUseAsyncs.useAsync-1": {
          "data": null,
          "depends": null,
          "error": null,
          "load_state": "loading",
        },
        "__cache": {},
      }
    `);

    const event = {
      asyncResponse: {
        requestId: (async1Ref.id ?? '') + '-null',
        data: { value: 'foo' },
        error: undefined,
      },
      hook: async1Ref.id,
    };

    const request2: UIRequest = { events: [event], state: response1.state };
    const response2 = await handler.handle(request2, mockMetadata);

    expect(response2.events).toMatchInlineSnapshot(`
      [
        {
          "async": true,
          "asyncRequest": {
            "requestId": "AppUseAsyncs.useAsync-1-"foo"",
          },
          "hook": "AppUseAsyncs.useAsync-1",
        },
      ]
    `);

    expect(response2.state).toMatchInlineSnapshot(`
      {
        "AppUseAsyncs.useAsync-0": {
          "data": "foo",
          "depends": null,
          "error": null,
          "load_state": "loaded",
        },
        "AppUseAsyncs.useAsync-1": {
          "data": null,
          "depends": "foo",
          "error": null,
          "load_state": "loading",
        },
        "__cache": {},
      }
    `);

    const event3 = {
      asyncResponse: {
        requestId: (async2Ref.id ?? '') + '-"foo"',
        data: { value: 'bar' },
        error: undefined,
      },
      hook: async2Ref.id,
    };

    const request3: UIRequest = { events: [event3], state: response2.state };
    const response3 = await handler.handle(request3, mockMetadata);

    expect(response3.events).toMatchInlineSnapshot(`[]`);
    expect(response3.state).toMatchInlineSnapshot(`
      {
        "AppUseAsyncs.useAsync-1": {
          "data": "bar",
          "depends": "foo",
          "error": null,
          "load_state": "loaded",
        },
        "__cache": {},
      }
    `);
  });
});

describe('circuit breaking', () => {
  test('circuit breaking doesnt return the error, it breaks', async () => {
    shouldThrow = true;
    const handler = new BlocksHandler(App);
    let request: UIRequest = EmptyRequest;
    const response = await handler.handle(request, mockMetadata);

    request = { events: [asyncRequestEvent(asyncRef)], state: response.state };
    await expect(async () => handler.handle(request, mockMetadata)).rejects.toThrowError(
      'ServerCallRequired'
    );
  });
});

describe('invalidation', () => {
  test('changing ids', async () => {
    const handler = new BlocksHandler(App);
    let request: UIRequest = EmptyRequest;
    let response = await handler.handle(request, mockMetadata);
    expect(response.state).toMatchInlineSnapshot(`
      {
        "App.useAsync-1": {
          "data": null,
          "depends": 1,
          "error": null,
          "load_state": "loading",
        },
        "App.useState-0": {
          "error": null,
          "load_state": "loaded",
          "value": 1,
        },
        "__cache": {},
      }
    `);
    const initialState = response.state;

    // ID unchanged, no new events emitted
    request = { events: [], state: initialState };
    response = await handler.handle(request, mockMetadata);
    expect(response.events.length).toEqual(0);

    //ID changed, new event emitted
    const altered = { ...initialState };
    altered['App.useState-0'] = { value: 'other' };
    request = { events: [], state: altered };
    response = await handler.handle(request, mockMetadata);
    expect(response.events).toEqual([asyncRequestEvent(asyncRef, 'other')]);

    // //ID changed, state also changed, no new event emitted
    // request = { events: [], state: response.state, props: { depends: 1 } };
    // response = await handler.handle(request, mockMetadata);
    // expect(response.events).toEqual([]);
  });

  test('mismatched responses', async () => {
    const handler = new BlocksHandler(App);
    let request: UIRequest = EmptyRequest;
    let response = await handler.handle(request, mockMetadata);
    const initialState = response.state;

    request = { events: [asyncResponseEvent(asyncRef, 'unknown')], state: initialState };
    response = await handler.handle(request, mockMetadata);
    expect(response.state).toEqual({ __cache: {} }); // nothing changes

    // This one is actually matched up.
    request = { events: [asyncResponseEvent(asyncRef)], state: initialState };
    response = await handler.handle(request, mockMetadata);
    expect(response.state).toMatchInlineSnapshot(`
      {
        "App.useAsync-1": {
          "data": "hi world!",
          "depends": 1,
          "error": null,
          "load_state": "loaded",
        },
        "__cache": {},
      }
    `);
  });
});

describe.each([true, false])(`when circuit-breaking: %s`, (circuitBreaking) => {
  test('initial load', async () => {
    shouldThrow = circuitBreaking;
    const handler = new BlocksHandler(App);

    const response = await handler.handle(EmptyRequest, mockMetadata);

    expect(JSON.stringify(response.blocks)).toContain('loading');
    expect(response.events.length).toEqual(1);
    expect(response.events[0]).toEqual(asyncRequestEvent(asyncRef));
  });

  test('receiving request', async () => {
    shouldThrow = false;
    const handler = new BlocksHandler(App);

    let response = await handler.handle(EmptyRequest, mockMetadata);

    const request: UIRequest = { events: [asyncRequestEvent(asyncRef)], state: response.state };
    expect(response.state).toMatchSnapshot();
    response = await handler.handle(request, mockMetadata);

    expect(response.blocks).toBeFalsy();
    expect(response.events.length).toEqual(1);
    expect(response.events[0]).toEqual(asyncResponseEvent(asyncRef));
  });

  test('receiving response', async () => {
    shouldThrow = true;
    const handler = new BlocksHandler(App);

    let response = await handler.handle(EmptyRequest, mockMetadata);
    expect(JSON.stringify(response.blocks)).toContain('loading');

    const request: UIRequest = { events: [asyncResponseEvent(asyncRef)], state: response.state };
    response = await handler.handle(request, mockMetadata);

    expect(findHookState(asyncRef)).toMatchSnapshot();
    expect(JSON.stringify(response.blocks)).toContain('hi world!');
    expect(response.effects.length).toEqual(0);
  });
});
