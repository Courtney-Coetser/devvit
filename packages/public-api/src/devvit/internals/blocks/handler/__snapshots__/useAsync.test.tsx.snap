// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`regressions > loading null doesnt loop 3`] = `
{
  "data": null,
  "depends": 1,
  "error": null,
  "load_state": "loaded",
}
`;

exports[`when circuit-breaking: false > receiving request 1`] = `
{
  "App.useAsync-2": {
    "data": null,
    "depends": 1,
    "error": null,
    "load_state": "loading",
  },
  "App.useState-0": 1,
  "App.useState-1": true,
}
`;

exports[`when circuit-breaking: false > receiving response 1`] = `
{
  "data": "hi world!",
  "depends": 1,
  "error": null,
  "load_state": "loaded",
}
`;

exports[`when circuit-breaking: true > receiving request 1`] = `
{
  "App.useAsync-2": {
    "data": null,
    "depends": 1,
    "error": null,
    "load_state": "loading",
  },
  "App.useState-0": 1,
  "App.useState-1": true,
}
`;

exports[`when circuit-breaking: true > receiving response 1`] = `
{
  "data": "hi world!",
  "depends": 1,
  "error": null,
  "load_state": "loaded",
}
`;
